org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/javadoc/JavaDocHTMLPathHandler.java

```diff
@@ -0,0 +1,308 @@
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.FileTime;
+import java.util.jar.JarFile;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.zip.ZipEntry;
+
+import org.eclipse.core.resources.IResource;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IPackageFragment;
+import org.eclipse.jdt.core.dom.TextElement;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+
+/**
+ * @author Nikolas Komonen - nkomonen@redhat.com
+ *
+ */
+public class JavaDocHTMLPathHandler {
+
+	// absolute path to folder holding all extracted images
+	public static final IPath EXTRACTED_JAR_IMAGES_FOLDER = JavaLanguageServerPlugin.getInstance().getStateLocation().append("extracted-jar-images/");
+
+	public static final String[] tags = { "img" };
+
+	public static final Pattern p = Pattern.compile("(src\\s*=\\s*['\"])");
+
+	/**
+	 * Returns true if the text is an HTML tag in the defined array of tag names in
+	 * {@link JavaDocHTMLPathHandler#tags}
+	 *
+	 * @param text
+	 * @return
+	 */
+	public static boolean containsHTMLTag(String text) {
+
+		for (String tag : tags) {
+			if (text.contains("<" + tag)) {
+				return true;
+			}
+		}
+
+		return false;
+	}
+
+	/**
+	 * Given a {@link TextElement} that represents an HTML tag with a 'src'
+	 * attribute, it will extract the image from the jar if necessary and copy it to
+	 * the 'outputPath'.
+	 *
+	 * @param child
+	 * @param fElement
+	 * @return
+	 */
+	public static String getValidatedHTMLSrcAttribute(TextElement child, IJavaElement fElement) {
+
+		//Check if current src attribute path needs to be validated
+		String text = child.getText();
+		int offsets[] = extractSourcePathFromHTMLTag(text);
+		if (offsets == null) {
+			return text;
+		}
+		String srcPath = text.substring(offsets[0], offsets[1]);
+
+		if (isPathAbsolute(srcPath)) {
+			return text; //Current path is good as is.
+		}
+
+		String fileName = Paths.get(srcPath).getFileName().toString();
+
+		//Get the initial internal jar fragment, which is going to be the relative parent folder to the file
+		IPackageFragment internalJarFragment = (IPackageFragment) fElement.getAncestor(IJavaElement.PACKAGE_FRAGMENT);
+
+		if (internalJarFragment == null) {
+			return text;
+		}
+
+		//test if srcPath points to the correct file, relative to the fElement file.
+		IResource fragmentResource = internalJarFragment.getResource();
+		if (fragmentResource != null) {
+			IPath absoluteFragmentPath = fragmentResource.getLocation();
+			File absoluteFragmentFile = new File(absoluteFragmentPath.toPortableString() + "/" + srcPath);
+			if (absoluteFragmentFile.exists()) {
+				String uriString = "file://" + absoluteFragmentFile.toURI().getPath();
```

Comment:
why not absoluteFragmentFile.toURI().toString()?

Reply:
ResourceUtils.fixURI(absoluteFragmentFile.toURI()).toString()